### ISSUE TITLE
'Unused local variables and functions should be removed' reported for variable which is used

### ISSUE THREAD

**Issue by rhiadj** [SonarSource/SonarJS#1951]
**I want to report a bug.**

**SonarJS version:** 6.2 (build 12043)<br>
**SonarQube version:** Community Edition Version 7.9.2 (build 30863)<br>

**Rule key:** S1481

**Reproducer** 
In a CRA project (built using Material UI) we have the following code:
```
const IconLabelButton = ({ icon, label, clickHandler }) => {
  const ButtonIcon = icon;
  return (
    <Button onClick={clickHandler}>
      <ButtonIcon />
      {label}
    </Button>
  );
};
```

**Logs**
`Remove the declaration of the unused 'ButtonIcon' variable.`
And...
`Unused local variables and functions should be removed`

**Expected behavior**
Whilst I appreciate the code could be rewritten to cast the variable name in the function declaration:
```
const IconLabelButton = ({ icon: ButtonIcon, label, clickHandler }) => {
```
... the use case above is valid/the error reported is misleading and causing minor issues to be reported for a supposed 'Unused local variable' (`ButtonIcon`) which is in fact used.

This was pre-existing code within our application - and was previously not reported - which started to be flagged after we updated SonarJS recently. Looking at recent updates I suspect these are the changes causing it?
https://github.com/SonarSource/SonarJS/pull/1910
https://github.com/SonarSource/SonarJS/issues/1905

**Small thank you...**
Many thanks for all your efforts! We use this tooling and the SonarJS plugin heavily on all of our projects and this is the first time I've had cause to raise a bug - so thank you!

**Comment by vilchik-elena** [comment]
Hi,

I managed to reproduce FP, thanks for reporting! And thanks for kind words :)

I don't understand what is the syntax used in 
```
const IconLabelButton = ({ icon: ButtonIcon, label, clickHandler }) => {
```

Is it TS?

**Comment by rhiadj** [comment]
Wow - gosh that was a very fast reply! Thanks for that - glad you were able to reproduce.

Re the code syntax - that's renaming a destructured variable:
https://codeburst.io/renaming-destructured-variables-in-es6-807549754972

Standard ES6 stuff but isn't very widely used AFAIK.

Thanks again! 👍

**Comment by vilchik-elena** [comment]
ah, indeed, I missed that it's not just parameters, but destructuring

**Comment by wellyal** [comment]
Hi, I'm still having a similar issue, I don't know if I am missing some sort of configuration to let the analyzer knows that I'm running react with a jsx sintax.

![image](https://user-images.githubusercontent.com/3047016/92483811-a6572d00-f1bf-11ea-9965-5820ba5df4bd.png)

**Comment by saberduck** [comment]
hello @wellyal , can you please create thread on our community forum https://community.sonarsource.com/c/bug/fp/7 ? It will be easier for us to help you there. Also can you please provide additional context, like do you use SonarCloud or SonarQube? If SonarQube, which version, etc...

**Comment by wellyal** [comment]
@saberduck hi, I did already. No one has answered me though.

https://community.sonarsource.com/t/rule-for-shadowed-variables-is-incorrect-for-javascript/30810

**Comment by agam-armis** [comment]
+1 on this still happening for `React` variable on JSX files

**Comment by vilchik-elena** [comment]
@agam-armis are you using sonarqube >=9.1? If yes, please create a new thread in community forum https://community.sonarsource.com/c/bug/fp/7 with a full reproducer
thanks!
